### PR TITLE
add backbone nets

### PR DESCRIPTION
**Pull request by vladiprof**

### CODE REVIEW

File: basenets/convnet.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from tensorflow.keras import layers, models, activations, backend
+
+
+class ConvBlock(layers.Layer):
+    def __init__(self,  out_channels, add_maxpool = True, **kwargs):
```
Comment (konev-artem) [inline basenets/convnet.py]: pep8 не пропустил бы "add_maxpool = True" с пробелами

File: basenets/convnet.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        if self.maxpool is not None:
+            out = self.maxpool(out)
+        
+        return out
+
+    def set_trainable(self, trainable):
+        self.conv.trainable = trainable
+        self.bn.trainable = trainable
+
+    def compute_output_shape(self, input_shape):
```
Comment (konev-artem) [inline basenets/convnet.py]: а что оно возвращает (что в input_shape[0])?

File: basenets/convnet.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        return (input_shape[0], self.out_channels)
+
+
+class ConvNet:
```
Comment (konev-artem) [inline basenets/convnet.py]: было бы полезно в комментариях написать ссылку на статью, откуда ты взял эту конфигурацию (Conv4)

File: basenets/resnet12.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+
+def swish1(x):
+    return x * activations.sigmoid(x)
+
+class ResidualBlock(layers.Layer):
+    def __init__(self,  out_channels, **kwargs):
+        self.conv1 = layers.Conv2D(out_channels, 3, padding='same')
+        self.bn1 = layers.BatchNormalization()
+        self.nl1 = layers.Lambda(swish1)
```
Comment (konev-artem) [inline basenets/resnet12.py]: в нашем resnet разве используется swish?
если да, то тут тоже нужна ссылка на статью, где это написано

File: tests/convnet4.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+from basenets.convnet import ConvNet
+
+
+class TestConvnet(unittest.TestCase):
+
+    def test_forward(self):
+        net = ConvNet()
+        model = models.Model(*net.build_net((256, 256, 3)))
+
+        np.random.seed(4412442)
```
Comment (konev-artem) [inline tests/convnet4.py]: я бы не ставил seed в тестах, где он не нужен

File: basenets/convnet.py (hunk 1 above)
Comment (vladiprof) [reply basenets/convnet.py]: fix

File: basenets/convnet.py (hunk 2 above)
Comment (vladiprof) [reply basenets/convnet.py]: починил, теперь возвращает правильные размеры выходного тензора

File: tests/convnet4.py (hunk 5 above)
Comment (vladiprof) [reply tests/convnet4.py]: убрал

File: basenets/convnet.py (hunk 3 above)
Comment (vladiprof) [reply basenets/convnet.py]: Добавил комментарий
"A Closer Look at Few-Shot Classification" Wei-Yu Chen.
за основу взял https://github.com/wyharveychen/CloserLookFewShot/blob/master/backbone.py

File: basenets/resnet12.py (hunk 4 above)
Comment (vladiprof) [reply basenets/resnet12.py]: Добавил комментарий. Статья: "Dense Classification and Implanting for Few-Shot Learning" https://arxiv.org/pdf/1903.05050.pdf

File: basenets/resnet12.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        self.max_pool1 = layers.MaxPool2D(padding='same')
+        self.residual2 = ResidualBlock(128)
+        self.max_pool2 = layers.MaxPool2D(padding='same')
+        self.residual3 = ResidualBlock(256)
+        self.max_pool3 = layers.MaxPool2D(padding='same')
+        self.residual4 = ResidualBlock(512)
+        self.max_pool4 = layers.MaxPool2D(padding='same')
+        
+        
+    def build_net(self, input_size):
```
Comment (konev-artem) [inline basenets/resnet12.py]: сеть создаем в конструкторе

File: basenets/resnet12.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        input = layers.Input(shape=input_size)
+        x = self.residual1(input)
+        x = self.max_pool1(x)
+        x = self.residual2(x)
+        x = self.max_pool2(x)
+        x = self.residual3(x)
+        x = self.max_pool3(x)
+        x = self.residual4(x)
+        x = self.max_pool4(x)       
+        return [input], [x]
```
Comment (konev-artem) [inline basenets/resnet12.py]: делаем ф-ции get_inputs/get_outputs

File: basenets/resnet12.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+def swish1(x):
+    return x * activations.sigmoid(x)
+
+
+# Resnet-12 from "Dense Classsification and Implanting for Few-Shot Learning" Yann Lifchitz...
+class ResidualBlock(layers.Layer):
+    def __init__(self,  out_channels, **kwargs):
+        self.conv1 = layers.Conv2D(out_channels, 3, padding='same')
+        self.bn1 = layers.BatchNormalization()
+        self.nl1 = layers.Lambda(swish1)
```
Comment (konev-artem) [inline basenets/resnet12.py]: сделать гибкую настройку типа ф-ции активации

File: basenets/resnet12.py (hunk 7 above)
Comment (vladiprof) [reply basenets/resnet12.py]: сделал

File: basenets/resnet12.py (hunk 6 above)
Comment (vladiprof) [reply basenets/resnet12.py]: сделал

File: fewshot/backbones/resnet12.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+class Resnet12:
+    def __init__(self, input_size, activation='swish1'):
+        self.residual1 = ResidualBlock(64, activation)
+        self.max_pool1 = layers.MaxPool2D(padding='same')
+        self.residual2 = ResidualBlock(128, activation)
+        self.max_pool2 = layers.MaxPool2D(padding='same')
+        self.residual3 = ResidualBlock(256, activation)
+        self.max_pool3 = layers.MaxPool2D(padding='same')
+        self.residual4 = ResidualBlock(512, activation)
+        self.max_pool4 = layers.MaxPool2D(padding='same')
```
Comment (highrut) [inline fewshot/backbones/resnet12.py]: мб, добавить maxpool в конец residual block, раз уж они только в такой последовательности встречаются?

File: fewshot/backbones/resnet12.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        self.out_channels = out_channels
+        super(ResidualBlock, self).__init__(**kwargs)
+
+    def create_activation(self, activation):
+        if activation == 'relu':
+            return layers.ReLU()
+        elif activation == 'swish1':
+            return layers.Lambda(swish1)
+        else:
+            raise NotImplemented('activation ' + activation + ' not found')
```
Comment (highrut) [inline fewshot/backbones/resnet12.py]: можно заменить слой ReLU на Activation -- и не ограничиваться этими двумя активациями. 
список активаций, который можно передавать строкой: https://keras.io/activations/

File: fewshot/backbones/convnet.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            out_shape = self.maxpool.compute_output_shape(out_shape)
+
+        return out_shape
+
+
+class ConvNet:
+    def __init__(self, input_size, depth=4):
+        self.blocks = []
+        for i in range(depth):
+            outdim = 64
```
Comment (highrut) [inline fewshot/backbones/convnet.py]: зачем создавать переменную с одним и тем же значением каждый раз в цикле?

File: fewshot/backbones/convnet.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
+from tensorflow.keras import layers, models, activations, backend
+
+
+# ConvNet-N from "A Closer Look at Few-Shot Classification" Wei-Yu Chen.
+# reference implementation: https://github.com/wyharveychen/CloserLookFewShot/blob/master/backbone.py
+class ConvBlock(layers.Layer):
+    def __init__(self,  out_channels, add_maxpool=True, **kwargs):
```
Comment (highrut) [inline fewshot/backbones/convnet.py]: пробел перед out

File: fewshot/backbones/resnet12.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        self.bn1 = layers.BatchNormalization()
+        self.nl1 = self.create_activation(activation)
+        
+        self.conv2 = layers.Conv2D(out_channels, 3, padding='same')
+        self.bn2 = layers.BatchNormalization()
+        self.nl2 = self.create_activation(activation)
+        
+        self.conv3 = layers.Conv2D(out_channels, 3, padding='same')
+        self.bn3 = layers.BatchNormalization()
+        self.nl3 = self.create_activation(activation)
```
Comment (highrut) [inline fewshot/backbones/resnet12.py]: conv2d + batchnorm + activation уже реализовано в виде ConvBlock -- можно использовать его и сократить количество строк в три раза

File: basenets/resnet12.py (hunk 8 above)
Comment (vladiprof) [reply basenets/resnet12.py]: сделал

File: fewshot/backbones/resnet12.py (hunk 9 above)
Comment (vladiprof) [reply fewshot/backbones/resnet12.py]: исправил

File: fewshot/backbones/resnet12.py (hunk 10 above)
Comment (vladiprof) [reply fewshot/backbones/resnet12.py]: сделал

File: fewshot/backbones/convnet.py (hunk 11 above)
Comment (vladiprof) [reply fewshot/backbones/convnet.py]: поправил

File: fewshot/backbones/convnet.py (hunk 12 above)
Comment (vladiprof) [reply fewshot/backbones/convnet.py]: поправил

File: fewshot/backbones/resnet12.py (hunk 13 above)
Comment (vladiprof) [reply fewshot/backbones/resnet12.py]: сделал

File: fewshot/backbones/resnet12.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
+from tensorflow.keras import layers, models, activations, backend
+import tensorflow.keras as K
+import fewshot.backbones.convnet as convnet
+
+
+# Resnet-12 from "Dense Classsification and Implanting for Few-Shot Learning" Yann Lifchitz...
+class ResidualBlock(layers.Layer):
+    def __init__(self,  out_channels, activation='swish1', **kwargs):
+        self.conv1 = convnet.ConvBlock(out_channels, activation, add_maxpool=False)
```
Comment (konev-artem) [inline fewshot/backbones/resnet12.py:9]: из статьи:
In most experiments we use a ResNet-12 net- work [31] as our embedding network, composed of four residual blocks [11], each having three 3×3 convolutional layers with batch normalization [16] and swish-1 activa- tion function [34]. Each block is followed by 2×2 max- pooling

почему add_maxpool=False?

File: fewshot/backbones/resnet12.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        self.residual4 = ResidualBlock(512, activation)
+        self.inputs, self.outputs = self._build_net(input_size)
+        
+    def _build_net(self, input_size):
+        input = layers.Input(shape=input_size)
+        x = self.residual1(input)
+        x = self.residual2(x)
+        x = self.residual3(x)
+        x = self.residual4(x)
+        return [input], [x]
```
Comment (highrut) [inline fewshot/backbones/resnet12.py:62]: зачем нужно возвращать листы, а не просто input, x?

File: fewshot/backbones/convnet.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        if self.maxpool is not None:
+            out_shape = self.maxpool.compute_output_shape(out_shape)
+
+        return out_shape
+
+
+class ConvNet:
+    def __init__(self, input_size, depth=4):
+        self.blocks = []
+        outdim = 64
```
Comment (konev-artem) [inline fewshot/backbones/convnet.py]: тоже можно было бы в параметры вынести

File: fewshot/backbones/resnet12.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        y = self.conv2(y)
+        y = self.conv3.conv(y)
+        y = self.conv3.bn(y)
+
+        z = self.conv_res(x)
+        z = self.bn_res(z)
+        
+        out = y + z
+        out = self.conv3.nl(out)
+        out = self.conv3.maxpool(out)
```
Comment (highrut) [inline fewshot/backbones/resnet12.py]: лучше не обращаться к членам объекта класса ConvBlock напрямую -- если у него как-то изменится внутренняя структура, то ResidualBlock сломается -- а, по-хорошему, не должен.

сейчас чтобы применить пулинг, ты создаешь его как элемент последнего ConvBlock. вместо этого можно перенести активацию и пулинг непосредственно в ResidualBlock, а conv3 создать с параметрами activation='linear', add_maxpool=False.

File: fewshot/backbones/resnet12.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+        self.conv_res  = layers.Conv2D(out_channels, 3, padding='same')
+        self.bn_res = layers.BatchNormalization()
+                
+        self.out_channels = out_channels
+        super(ResidualBlock, self).__init__(**kwargs)
+
+    def call(self, x):
+        y = self.conv1(x)
+        y = self.conv2(y)
+        y = self.conv3.conv(y)
```
Comment (konev-artem) [inline fewshot/backbones/resnet12.py]: лучше сделать conv3 с activation=None и не делать тут все эти строчки с conv3.*

File: fewshot/backbones/resnet12.py (hunk 14 above)
Comment (vladiprof) [reply fewshot/backbones/resnet12.py:9]: Residual  блок состоит из 3-х conv + bn + swish, и после блока стоит maxpool. Те для первых трех сверток не неужен макспул

File: fewshot/backbones/convnet.py (hunk 16 above)
Comment (vladiprof) [reply fewshot/backbones/convnet.py]: сделал

File: fewshot/backbones/resnet12.py (hunk 18 above)
Comment (vladiprof) [reply fewshot/backbones/resnet12.py]: сделал

File: fewshot/backbones/resnet12.py (hunk 17 above)
Comment (vladiprof) [reply fewshot/backbones/resnet12.py]: поправил

File: fewshot/backbones/resnet12.py (hunk 15 above)
Comment (vladiprof) [reply fewshot/backbones/resnet12.py:62]: потенциально и умодели может быть несколько входов/выходов

File: fewshot/backbones/resnet12.py (hunk 14 above)
Comment (konev-artem) [reply fewshot/backbones/resnet12.py:9]: А, значит расположение ссылки на статью меня запутало - я ожидал, что этот класс реализует resnet12
